### PR TITLE
Use PK of Descendants in TreeNodeMultipleChoiceFilter

### PR DESCRIPTION
**Pull request by bryanculver**
# Closes: #4329
# What's Changed

- Used the `pk` of the descendants instead of `name` which is non-unique. This _could_ be backported to 1.6 but we used `slug` there which was globally unique.

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [NA] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [NA] Documentation Updates (when adding/changing features)
- [NA] Example Plugin Updates (when adding/changing features)
- [NA] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by bryanculver** [comment]
Using the dataset constructed in #4329, fixed: 
image

File: nautobot/core/filters.py [hunk 1]
```diff
@@ -501,7 +501,7 @@ def generate_query(self, value, qs=None, **kwargs):
         predicates = []
         for obj in value:
             # Try to get the `to_field_name` (e.g. `name`) or just pass the object through.
-            val = getattr(obj, self.field.to_field_name, obj)
```
Comment (HanlinMiao) [inline nautobot/core/filters.py:504]: For my own edification, how is the filter still able to filter on `name` if we make this change here?

Comment (bryanculver) [reply nautobot/core/filters.py:504]: @glennmatthews Might be able to point to the line where this is happening but by the time we get to `generate_query` we already have actual instances (hence the `node.descendants(include_self=True)` above). When you intend to search for something by name, you want to find all instances that match that name. Without this fix, it is using the name of the descendants to then perform the query again (find all instances that match the name of my descendants) you expand the resulting query.

Something like:

```
Blues
  Teal
  Aquamarine
  Sky
Reds
  Crimson
  Strawberry
  Burgundy
```

If you search for "Devices in Locations with the name 'Red'", the resulting filter becomes "Devices in Locations with the names ['Red', 'Crimson', 'Strawberry', 'Burgundy']". Fine but if you add "Purple" as a Location in both Blues and Reds, Devices in Blues > Purple would show up too.

Instead what is performed is "Give me all Locations with the name 'Red'" and then to process the tree, it then performs "Give me all Locations matching this PK which are the descendants of Locations with the name 'Red'".

File: nautobot/core/filters.py [hunk 2]
```diff
@@ -501,7 +501,7 @@ def generate_query(self, value, qs=None, **kwargs):
         predicates = []
         for obj in value:
             # Try to get the `to_field_name` (e.g. `name`) or just pass the object through.
```
Comment (bryanculver) [inline nautobot/core/filters.py:503]:
```suggestion
            # Get the exact instance by PK as we are nested from the original query,
            #   or just pass the object through, commonly the null case.
```
@HanlinMiao?

File: nautobot/core/tests/test_filters.py [hunk 3]
```diff
@@ -87,26 +87,62 @@ def setUp(self):
...
+            status=status,
+        )
         self.queryset = dcim_models.Location.objects.filter(name__icontains="Test Child Location")
 
     def test_filter_single_name(self):
         kwargs = {"parent": [self.parent_location_1.name]}
         qs = self.LocationFilterSet(kwargs, self.queryset).qs
 
-        self.assertQuerysetEqual(qs, [self.child_location_1])
+        self.assertQuerysetEqual(qs, [self.child_location_1, self.child_location_same_name_1])
```
Comment (glennmatthews) [inline nautobot/core/tests/test_filters.py:124]: Hmm, something seems off here - why isn't `self.parent_location_same_name_1` expected to be part of this result too? It's a child of `child_location_1` and a parent of `child_location_same_name_1` so why are they both returned but not it?

Comment (bryanculver) [reply nautobot/core/tests/test_filters.py:124]: Note the name filter of the base QuerySet.

Comment (HanlinMiao) [reply nautobot/core/tests/test_filters.py:122]: This line 90/118 filters out all the locations with name "Test Parent Location" etc.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**